### DISCUSSION TITLE
Does curl send a FIN to the destination host on exit?

### DISCUSSION THREAD

**Discussion by ksudate** [curl/curl#8874] [Q&A] [answered]
Hello.

I was using curl to check the operation of http keep-alive.
```
$ curl <destination server> -I
HTTP/1.1 200 OK
Server: nginx/1.18.0 (Ubuntu)
Date: Wed, 18 May 2022 06:07:56 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Mon, 16 May 2022 06:27:59 GMT
Connection: keep-alive
ETag: "6281eeef-264"
Accept-Ranges: bytes
```
`Connection: keep-alive` is granted, but actually a FIN is sent and the Connection is closed.
We confirmed with tcpdump that curl sends FIN even when http keep-alive is enabled.
When `--next` option was used, it still sent a FIN on the last request.
(The first request did not send a FIN, confirming that HTTP Keep-Alive is enabled.)
```
$ curl <destination server> -I --next <destination server> -I
HTTP/1.1 200 OK
Server: nginx/1.18.0 (Ubuntu)
Date: Wed, 18 May 2022 05:44:49 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Mon, 16 May 2022 06:27:59 GMT
Connection: keep-alive
ETag: "6281eeef-264"
Accept-Ranges: bytes

HTTP/1.1 200 OK
Server: nginx/1.18.0 (Ubuntu)
Date: Wed, 18 May 2022 05:44:49 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Mon, 16 May 2022 06:27:59 GMT
Connection: keep-alive
ETag: "6281eeef-264"
Accept-Ranges: bytes
```

Does curl send a FIN to the destination host on exit?

Thank you very much for any help.

**Comment by jzakrzewski** [comment, marked as answer]
I'm guessing a bit here but I think that FIN will be send by the underlying TCP/IP implementation (the OS) when closing the socket (closing the socket breaks the connection anyway, so it makes sense to do it properly). And the socket will be closed on program exit by the OS if not done explicitly earlier.

**Reply by ksudate** [reply]
Thanks for the reply!
Then I understood that http keep-alive is only enabled when multiple requests are made in one curl run using `--next`.

**Reply by ksudate** [reply]
Just to add a little, I understood that http keep-alive is not enabled, but that multiple requests in one tcp connection using curl can only be achieved if `--next` is used.

**Reply by jzakrzewski** [reply]
Keep-Alive does not make sense if one side of the connection dies. If you had a long-lived process using libcurl, it'd be handled correctly but the curl tool just does the transfers and exits, so yeah, the connection will be kept alive as requested but only up until the process exits.

**Reply by ksudate** [reply]
I found your answers helpful, thank you very much!
